src/main/java/gov/usgs/earthquake/nats/NATSClient.java

```diff
@@ -0,0 +1,184 @@
+    this("localhost","4222","test-cluster",Long.toString(Thread.currentThread().getId()));
+  }
+
+  public NATSClient(String serverHost, String serverPort, String clusterId, String clientIdSuffix) {
+    // try to generate a unique ID; use suffix only if fail
+    this.serverHost = serverHost;
+    this.serverPort = serverPort;
+    this.clusterId = clusterId;
+    this.clientIdSuffix = clientIdSuffix;
+  }
+
+  /**
+   * Configures the required and optional parameters to connect to NATS Streaming server
+   *
+   * @param config
+   *            the Config to load.
+   * @throws Exception
+   */
+  @Override
+  public void configure(Config config) throws Exception {
+    // required parameters
+    serverHost = config.getProperty(SERVER_HOST_PROPERTY);
+    if (serverHost == null) {
+      throw new ConfigurationException(SERVER_HOST_PROPERTY + " is a required parameter");
+    }
+
+    serverPort = config.getProperty(SERVER_PORT_PROPERTY);
+    if (serverHost == null) {
+      throw new ConfigurationException(SERVER_PORT_PROPERTY + " is a required parameter");
+    }
+
+    clusterId = config.getProperty(CLUSTER_ID_PROPERTY);
+    if (serverHost == null) {
+      throw new ConfigurationException(CLUSTER_ID_PROPERTY + " is a required parameter");
+    }
+
+    clientId = config.getProperty(CLIENT_ID_PROPERTY);
+  }
+
+  /**
+   * Starts connection to NATS streaming server
+   * @throws Exception If something goes wrong when connecting to NATS streaming server
+   */
+  @Override
+  public void startup() throws Exception {
+    // generate client ID if we don't have one
+    if (clientId == null) {
+      clientId = generateClientId(clientIdSuffix);
+    }
+
+    // create connection
+    StreamingConnectionFactory factory = new StreamingConnectionFactory(clusterId,clientId);
+    factory.setNatsUrl("nats://" + serverHost + ":" + serverPort);
+    connection = factory.createConnection();
+  }
+
+  /**
+   * Safely closes connection to NATS Streaming server
+   */
+  @Override
+  public void shutdown() {
+    try {
+      connection.close();
+    } catch (Exception e) {
+      LOGGER.log(Level.WARNING, "[" + getName() + "] failed to close NATS connection");
+    }
+    connection = null;
+  }
+
+  /**
+   * Creates a client ID based on the host IP and MAC address
+   *
+   * @param suffix
+   *    Suffix to add to generated ID
+   * @return clientId
+   * @throws Exception if there's an issue accessing IP or MAC addresses, or can't do sha1 hash
+   */
+  private static String generateClientId(String suffix) throws Exception{
+    // get mac address
+    InetAddress host = InetAddress.getLocalHost();
+    NetworkInterface net = NetworkInterface.getByInetAddress(host);
+    byte[] macRaw =  net.getHardwareAddress();
+
+    // do a sha1 hash
+    MessageDigest digest = MessageDigest.getInstance("SHA-1");
+    digest.reset();
+    digest.update(macRaw);
+    String sha1 = String.format("%040x", new BigInteger(1, digest.digest()));
```

Comment:
This seems cryptic.  What about something like this base64 encoding?  https://github.com/usgs/pdl/blob/master/src/main/java/gov/usgs/earthquake/product/ProductDigest.java#L85

Reply:
It was cryptic, because I didn't really understand it. This is better.